### ISSUE TITLE
Sentiment train binaryModel fails

### ISSUE THREAD

**Issue by MalteFlender** [stanfordnlp/CoreNLP#406]
I am trying to train the Stanford NLP sentiment analysis tool with the default test and train set from https://nlp.stanford.edu/sentiment/trainDevTestTrees_PTB.zip in binary mode. I am using stanford-corenlp-full-2016-10-31.

To do so I use this command:

        java -cp "*" -Xmx10g edu.stanford.nlp.sentiment.SentimentTraining         
        -trainPath train.txt 
        -devPath dev.txt 
        -train 
        -model ORIG_BIN.ser.gz 
        -binaryModel

Which leads to the following error:

        [main] INFO edu.stanford.nlp.sentiment.SentimentTraining - Read in 8544 training trees
        [main] INFO edu.stanford.nlp.sentiment.SentimentTraining - Read in 1101 dev trees
        [main] INFO edu.stanford.nlp.sentiment.SentimentTraining - Sentiment model options:
        GENERAL OPTIONS
        randomSeed=-1081864962
        wordVectors=null
        unkWord=UNK
        randomWordVectors=true
        numHid=25
        numClasses=2
        lowercaseWordVectors=false
        useTensors=true
        simplifiedModel=true
        combineClassification=true
        classNames=Negative,Positive
        equivalenceClasses=0;1
        equivalenceClassNames=Negative,Positive
        TRAIN OPTIONS
        batchSize=27
        epochs=400
        debugOutputEpochs=8
        maxTrainTimeSeconds=86400
        learningRate=0.01
        scalingForInit=1.0
        classWeights=null
        regTransformMatrix=0.001
        regTransformTensor=0.001
        regClassification=1.0E-4
        regWordVector=1.0E-4
        initialAdagradWeight=0.0
        adagradResetFrequency=1
        shuffleMatrices=true
        initialMatrixLogPath=null
        nThreads=1
        TEST OPTIONS
        ngramRecordSize=0
        ngramRecordMaximumLength=0
        printLengthAccuracies=false
        [main] INFO edu.stanford.nlp.sentiment.SentimentTraining - Training on 8544 trees in 317 batches
        [main] INFO edu.stanford.nlp.sentiment.SentimentTraining - Times through each training batch: 400
        [main] INFO edu.stanford.nlp.sentiment.SentimentTraining - ======================================
        [main] INFO edu.stanford.nlp.sentiment.SentimentTraining - Starting epoch 0
        [main] INFO edu.stanford.nlp.sentiment.SentimentTraining - ======================================
        [main] INFO edu.stanford.nlp.sentiment.SentimentTraining - Epoch 0 batch 0
        Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 3
                at org.ejml.data.D1Matrix64F.set(Unknown Source)
                at org.ejml.simple.SimpleBase.set(Unknown Source)
                at edu.stanford.nlp.sentiment.SentimentCostAndGradient.backpropDerivativesAndError(SentimentCostAndGradient.java:374)
                at edu.stanford.nlp.sentiment.SentimentCostAndGradient.backpropDerivativesAndError(SentimentCostAndGradient.java:352)
                at edu.stanford.nlp.sentiment.SentimentCostAndGradient.scoreDerivatives(SentimentCostAndGradient.java:221)
                at edu.stanford.nlp.sentiment.SentimentCostAndGradient.calculate(SentimentCostAndGradient.java:247)
                at edu.stanford.nlp.optimization.AbstractCachingDiffFunction.ensure(AbstractCachingDiffFunction.java:140)
                at edu.stanford.nlp.optimization.AbstractCachingDiffFunction.derivativeAt(AbstractCachingDiffFunction.java:151)
                at edu.stanford.nlp.sentiment.SentimentTraining.executeOneTrainingBatch(SentimentTraining.java:33)
                at edu.stanford.nlp.sentiment.SentimentTraining.train(SentimentTraining.java:83)
                at edu.stanford.nlp.sentiment.SentimentTraining.main(SentimentTraining.java:230)

I also tried several other options like `-equivalenceClasses, -equivalenceClassNames -numClasses, -classNames` with more or less the same results.

What options do I need to make this work ?

----------

I am also interested into training my own corpus with three classes (Positive, Negative, Neutral) but with this task I had the same problems.

----------

I also (like suggested in https://nlp.stanford.edu/software/) made a stackoverflow question: http://stackoverflow.com/questions/43250625/stanford-nlp-sentiment-train-binarymodel-fails

**Comment by J38** [comment]
If you remove the `-binaryModel` the example will work.

I think the issue is that you are trying to train a binary model, but the example data is not binary but 5-class.

**Comment by MalteFlender** [comment]
Changing the input to 1 and 0 works.
But I thought that the `-equivalenceClasses` would tread 
0 and 1 (Very Negative and Negative) and 
3 and 4 (Positive and Very Positive) as [equivalence](https://github.com/stanfordnlp/CoreNLP/blob/master/src/edu/stanford/nlp/sentiment/RNNOptions.java#L77), so that changing the input data is not necessary.

If this is not the case, what dose the `-equivalenceClasses` option do and what do I have to use in order to train a 3 class (Positive, Neutral, Negative) model?
